uqcsbot/remindme.py

```diff
@@ -0,0 +1,408 @@
+DATETIME_VALID_FORMAT_ERR = "Dates and times should conform to a valid ISO 8601 format."
+DATETIME_IN_PAST_ERR = "Datetime can't be in the past."
+STARTEND_DATE_IN_PAST_ERR = "Start or end date can't be in the past."
+END_DATE_BEFORE_START_ERR = "End date can't be before the start date."
+NEGATIVE_WEEK_FREQUENCY_ERR = "Week frequency must be non-negative."
+
+REMOVE_REMINDERS_FOOTER = "You can remove active reminders with /remindme remove."
+LIST_REMINDERS_FOOTER = "You can view your active reminders with /remindme list."
+
+DISPLAY_DATE_FORMAT = '%A %d %b %Y'
+DISPLAY_DAY_NAME_FORMAT = '%A'
+DISPLAY_TIME_FORMAT = '%-I:%M %p'
+
+
+class Reminder(NamedTuple):
+    id: int
+    user_id: int
+    channel_id: int | None  # None => send reminder in DMs
+    time_created: dt.datetime
+    message: str
+    time: dt.time
+    start_date: dt.date
+    end_date: dt.date | None  # None => non-ending
+    week_frequency: int | None  # 0 => daily, None => one-time
+
+    def __str__(self):
+        id = self.id
+        message = self.message
+        time = self.time.strftime(DISPLAY_TIME_FORMAT)
+        start_date = self.start_date.strftime(DISPLAY_DATE_FORMAT)
+
+        if self.week_frequency == None:  # one-time reminders
+            return f"Reminder with id {id} set for {start_date} at {time}:\n> {message}"
+
+        if self.week_frequency == 0:
+            day = "daily"
+            frequency = ""
+        else:  # convert day to word; Monday, Tuesday, etc.
+            day = self.start_date.strftime(DISPLAY_DAY_NAME_FORMAT) + "s" 
+            frequency = (", repeating every week" if self.week_frequency == 1 else
+                         f", repeating every {self.week_frequency} weeks")
+
+        if self.end_date == None:  # recurring reminders (non-ending)
+            return (f"Reminder with id {id} set for {day} at {time} starting {start_date}{frequency}:\n> {message}")
+        # recurring reminders (ending on self.end_date)
+        end_date = self.end_date.strftime(DISPLAY_DATE_FORMAT)
+        return (f"Reminder with id {id} set for {day} at {time} starting {start_date} and "
+                f"ending {end_date}{frequency}:\n> {message}")
+    
+    def removed_str(self):
+        return f"Reminder with id {self.id} removed:\n> {self.message}"
+
+def _error_embed(description: str, footer: str = "") -> discord.Embed:
+    """ Returns a formatted Error embed with the given description and an optional footer """
+    embed = discord.Embed(title=REMINDME_ERROR_TITLE, description=description, color=discord.Color.red())
+    if footer != "":
+        embed.set_footer(text=footer)
+    return embed
+
+def _add_reminder_embed(reminder: Reminder, footer: str = "") -> discord.Embed:
+    """ Returns a formatted Reminder Added embed with the given Reminder and an optional footer """
+    embed = discord.Embed(title=REMINDME_ADDED_TITLE, description=str(reminder))
```

Comment:
2310 moment: defining all these instead of just using them in the one place they crop up :joy: (not a problem, just memeing)